app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.toolbar
+
+import android.content.Context
+import android.view.View
+import android.widget.TextView
+import androidx.annotation.ColorRes
+import androidx.annotation.DrawableRes
+import androidx.core.content.ContextCompat
+import mozilla.components.browser.menu.BrowserMenu
+import mozilla.components.browser.menu.item.BrowserMenuImageText
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.components
+
+internal const val NO_ID = -1
+
+internal fun TextView.setColorResource(@ColorRes textColorResource: Int) {
+    if (textColorResource != NO_ID) {
+        setTextColor(ContextCompat.getColor(context, textColorResource))
+    }
+}
+
+/**
+ * A helper function to get the proper label for the sign in button.  There are three states:
+ * user is signed-in, please sign-in, and we are not sure what's happening. The last state does
+ * occur when the menu is opened before the AccountManager has finished updating its state, or
+ * when there is no connection and AccountManager doesn't return us the profile because its cache
+ * has 2 minutes 'freshness':
+ * https://searchfox.org/mozilla-mobile/source/application-services/components/fxa-client/src/internal/profile.rs#41
+ */
+fun Context.getSignInLabel(): String {
+    val isAuthenticated = components.backgroundServices.accountManager.authenticatedAccount() != null
+    val isAccountManagerReady = components.backgroundServices.accountManagerAvailableQueue.isReady()
+    val email = components.backgroundServices.accountManager.accountProfile()?.email
+
+    return if (isAccountManagerReady) {
+        if (isAuthenticated) {
+            // It sounds a bit counterintuitive, but authenticated state doesn't mean we
+            // have the profile information. The state can be (and usually is) restored
+            // without connection to the Internet, but profile can not. So we have to make
+            // sure that we actually have the email in order to show it.
+            email ?: resources.getString(R.string.browser_menu_account_settings)
+        } else {
+            getString(R.string.sync_menu_sync_and_save_data)
+        }
+    } else {
+        resources.getString(R.string.browser_menu_account_settings)
+    }
+}
```

Comment:
Some suggestions for changes:

Since we already have access to the components, we could use the `SyncStore` instead of the `accountManager`. I believe that moving forward we would prefer the `SyncStore` abstraction, both because it follows the `State` library pattern more closely and because it will allow us to begin refactoring some of the complexity in the underlying Sync components. This could be done in combination with the `Settings.signedInFxaAccount` property, something like
```
if (settings.signedInFxaAccount) {
    syncStore.state.account?.email ?: R.string.browser_menu_account_settings
} else {
    R.string.sync_menu_sync_and_save_data
}
```

Reply:
Ah, I see now this function is used in both the `DefaultToolbarMenu` and the `HomeMenu`. In that case, we clearly need to keep it. You could consider moving it into `ext/SyncState` instead, as that feels like a more common location between `home` and `components/toolbar` to me. Something like:

```
SyncState.getSignInLabel(settings: Settings) String {
    ...
}
```

This would have the added benefit of being more easily testable than an extension defined on `Context`.